CMake/FindEnzyme.cmake

```diff
@@ -0,0 +1,194 @@
+  ClangEnzyme-16.so
+  ClangEnzyme-15.so
+  ClangEnzyme-14.so
+  ClangEnzyme-13.so
+  ClangEnzyme-12.so
+  ClangEnzyme-11.so
+  ClangEnzyme-10.so
+  ClangEnzyme-9.so
+  PATHS
+  ${ENZYME_DIR} $ENV{ENZYME_DIR}
+  ENV LD_LIBRARY_PATH ENV DYLD_LIBRARY_PATH
+  PATH_SUFFIXES
+  lib64 lib)
+
+find_program(GRIDKIT_LLVM_LINK llvm-link)
+find_program(GRIDKIT_OPT opt)
+
+if("${GRIDKIT_LLVM_LINK}" STREQUAL "GRIDKIT_LLVM_LINK-NOTFOUND")
+  message(FATAL_ERROR "Could not find llvm-link! Will not be able to build Enzyme targets.")
+endif()
+
+if("${GRIDKIT_OPT}" STREQUAL "GRIDKIT_OPT-NOTFOUND")
+  message(FATAL_ERROR "Could not find opt! Will not be able to build Enzyme targets.")
+endif()
+
+message(STATUS "${ENZYME_CLANG_PLUGIN_LIBRARY};${ENZYME_LLVM_PLUGIN_LIBRARY}")
+if((NOT ${ENZYME_LLVM_PLUGIN_LIBRARY}) OR (NOT ${ENZYME_CLANG_PLUGIN_LIBRARY}))
+  set(ENZYME_CLANG_PLUGIN_LIBRARY CACHE FILEPATH "Path to Enzyme Clang plugin library")
+  set(ENZYME_LLVM_PLUGIN_LIBRARY CACHE FILEPATH "Path to Enzyme LLVM plugin library")
+  message(STATUS "Found Enzyme clang plugin: ${ENZYME_CLANG_PLUGIN_LIBRARY}")
+  message(STATUS "Found Enzyme LLVM plugin: ${ENZYME_LLVM_PLUGIN_LIBRARY}")
+  get_filename_component(ENZYME_LIBRARY_DIR ${ENZYME_CLANG_PLUGIN_LIBRARY} DIRECTORY CACHE "Enzyme library directory")
+else()
+  message(FATAL_ERROR "Enzyme library not found!"
+    " Set ENZYME_DIR to Enzyme's installation prefix.")
+endif()
+
+macro(enzyme_add_executable)
+  set(options)
+  set(oneValueArgs NAME)
+  set(multiValueArgs SOURCES LINK_LIBRARIES)
+  cmake_parse_arguments(enzyme_add_executable "${options}" "${oneValueArgs}"
+    "${multiValueArgs}" ${ARGN})
+
+  set(PHASE2 "${CMAKE_CURRENT_BINARY_DIR}/${enzyme_add_executable_NAME}.bc")
+  set(PHASE3 "${CMAKE_CURRENT_BINARY_DIR}/${enzyme_add_executable_NAME}_enzyme.ll")
+  set(PHASE4 "${CMAKE_CURRENT_BINARY_DIR}/${enzyme_add_executable_NAME}_opt.ll")
+  set(PHASE5 "${CMAKE_CURRENT_BINARY_DIR}/${enzyme_add_executable_NAME}")
+
+  set(OBJS "")
+
+  foreach(SRC ${enzyme_add_executable_SOURCES})
+    set(PHASE0 "${CMAKE_CURRENT_SOURCE_DIR}/${SRC}")
+    set(PHASE1 "${CMAKE_CURRENT_BINARY_DIR}/${enzyme_add_executable_NAME}_${SRC}_compile.o")
+    add_custom_command(
+      DEPENDS ${PHASE0}
+      OUTPUT ${PHASE1}
+      COMMAND ${CMAKE_CXX_COMPILER} -flto -c ${PHASE0} -O2 -fno-vectorize -ffast-math -fno-unroll-loops -o ${PHASE1}
+      COMMENT "Compiling ${SRC} to object file for target ${enzyme_add_executable_NAME}"
+      )
+    set(OBJS "${OBJS} ${PHASE1}")
+  endforeach()
+
+  cmake_language(EVAL CODE "
+  add_custom_command(
+    DEPENDS ${OBJS}
+    OUTPUT ${PHASE2}
+    COMMAND ${GRIDKIT_LLVM_LINK} ${OBJS} -o ${PHASE2}
+    COMMENT \"Linking object files to LLVM bytecode for target ${enzyme_add_executable_NAME}\"
+    )
+  ")
+
+  add_custom_command(
+    DEPENDS ${PHASE2}
+    OUTPUT ${PHASE3}
+    COMMAND ${GRIDKIT_OPT} ${PHASE2} -load-pass-plugin=${ENZYME_LLVM_PLUGIN_LIBRARY} -passes=enzyme -o ${PHASE3} -S
```

Comment:
If you don't want such a manual pipeline, you can use the plugin registration hooks that the opt pass plugin api provides (like `registerOptimizerLastEPCallback`, for example). [See the docs here](https://llvm.org/doxygen/classllvm_1_1PassBuilder.html). Sorry I left you with such cursed cmake! 😅  You could load the plugin and use a command like this:
```
clang++ -fpass-plugin=LLVMEnzyme.so -mllvm '-passes=default<O1>,enzyme,default<O3>' -o ./a.out
```

Reply:
So, in CMake configuration you would set something like `-DCMAKE_CXX_FLAGS="-fpass-plugin=LLVMEnzyme.so" -mllvm '-passes=default<O1>,enzyme,default<O3>'`?

Reply:
Exactly. Instead of dumping the IR between phases, you can first register the enzyme pass as a plugin and then use the `-mllvm -passes=` argument to build a custom phase ordering so enzyme is thrown in there.